### DISCUSSION TITLE
ChangeDetectionTask - how to implement

### DISCUSSION THREAD

**Discussion by robmarkcole** [torchgeo/torchgeo#1535] [Ideas]
As noted in issue https://github.com/microsoft/torchgeo/issues/336 there is not currently a `ChangeDetectionTask` I am seeking guidance on how to implement this task, or instruction on how `SegmentationTask` or similar can be utilised for change detection. 
Thanks in advance

**Comment by adamjstewart** [comment]
`SemanticSegmentationTask` could be utilized by stacking all images along the channel dimension: B x T x C x H x W -> B x (T x C) x H x W

I also wouldn't be opposed to having a dedicated `ChangeDetectionTask` that takes samples with "image1" and "image2" keys. We mostly just didn't settle on a naming scheme. I think it also depends on whether or not there are custom models that only make sense for change detection and require images to be separate.

**Reply by robmarkcole** [reply]
One model I am interested in is TinyCD [which takes pairs of images](https://github.com/AndreaCodegoni/Tiny_model_4_CD/blob/main/models/change_classifier.py) - there also might be situations where you would want to apply a transform on one of the pair but not the other (perhaps a seasonal augmentation). 

Looking at OSCD the [images are stacked](https://github.com/microsoft/torchgeo/blob/7a793f3437222136a0b4eba2fb1bc0fcd947f604/torchgeo/datasets/oscd.py#L133), so would need to be unstacked for tinycd

**Reply by adamjstewart** [reply]
In that case I think we should add a `ChangeDetectionTask` and unstack CD datasets.

**Reply by robmarkcole** [reply]
MVP will focus on binary change detection, and report metrics per class change/no-change

**Reply by isaaccorley** [reply]
You could make a wrapper for TinyCD which would unstack the images in the forward pass.

**Comment by isaaccorley** [comment]
The downside of moving to keys of `image1` and `image2` instead of the original stacked channels format is that we can no longer use the datasets with `SemanticSegmentationTask`. So either way a preprocessing transform would need to be defined to restack the images together vs using image1/image2 keys

**Reply by robmarkcole** [reply]
Explored a solution to this in https://github.com/microsoft/torchgeo/pull/1711

Additionally we could provide a custom transform to perform the stacking
